csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -83,25 +85,121 @@ private protected IList<ArraySegment<byte>> Encapsulation
+        /// doesn't contain an binary context it returns an empty list.</summary>
+        private protected IList<ArraySegment<byte>> BinaryContext
+        {
+            get
+            {
+                var binaryContext = new List<ArraySegment<byte>>();
+                if (_encapsulationEnd is OutputStream.Position encapsulationEnd &&
+                    _binaryContextEnd is OutputStream.Position binaryContextEnd)
+                {
+                    if (encapsulationEnd.Segment == binaryContextEnd.Segment)
+                    {
+                        binaryContext.Add(Payload[encapsulationEnd.Segment].Slice(
+                            encapsulationEnd.Offset,
+                            binaryContextEnd.Offset - encapsulationEnd.Offset));
+                    }
+                    else
+                    {
+                        ArraySegment<byte> segment = Payload[encapsulationEnd.Segment].Slice(
+                            encapsulationEnd.Offset);
+                        if (segment.Count > 0)
+                        {
+                            binaryContext.Add(segment);
+                        }
+                        for (int i = encapsulationEnd.Segment + 1; i < binaryContextEnd.Segment; i++)
+                        {
+                            binaryContext.Add(Payload[i]);
+                        }
+
+                        segment = Payload[binaryContextEnd.Segment].Slice(0, binaryContextEnd.Offset);
+                        if (segment.Count > 0)
+                        {
+                            binaryContext.Add(segment);
+                        }
+                    }
+                }
+                return binaryContext;
+            }
+        }
+
+        private readonly HashSet<int> _binaryContextKeys = new HashSet<int>();
         private readonly CompressionLevel _compressionLevel;
         private readonly int _compressionMinSize;
 
         private IList<ArraySegment<byte>>? _encapsulation;
 
+        /// <summary>Write a binary context entry to the frame with the given key and value, the value memory is copied
+        /// to the frame internal buffer.</summary>
+        /// <param name="key">The binary context entry key.</param>
+        /// <param name="value">The binary context entry value.</param>
+        /// <exception cref="NotSupportedException">If the frame protocol doesn't support binary context.</exception>
+        /// <exception cref="InvalidOperationException">If the frame encapsulation has not been written.</exception>
+        /// <exception cref="ArgumentException">If the key is already in use.</exception>
+        public void AddBinaryContextEntry(int key, ReadOnlySpan<byte> value)
+        {
+            OutputStream ostr = WriteBinaryContextEntryKey(key);
+            Span<byte> size = stackalloc byte[2];
+            OutputStream.WriteFixedLengthSize20(value.Length, size);
```

Comment:
There is no need to create an OutputStream here, and you can write this known size immediately with WriteSize20 - and it could be any number of bytes.

Reply:
We can write the size without creating an OutputStream but we will have to deal with expanding the array of segments, maybe we can add a static Expand

Reply:
Fixed the code to cache the OutputStream used to write the binary context